docs/configuration.md
# Configuration

Piston provides many different configuration options to tweak Piston to meet your needs.

Configuration is specified through environment variables, prefixed with `PISTON_`.

## Log Level

```yaml
key: PISTON_LOG_LEVEL
default: INFO
```

Level of log output to provide.

One of `DEBUG`, `INFO`, `WARN`, `ERROR` or `NONE`

## Bind Address

```yaml
key: PISTON_BIND_ADDRESS
default: 0.0.0.0:2000
```

Port and IP address to bind the Piston API to.

<!-- prettier-ignore -->
!!! warning
    Changing this value is not recommended.

    This changes the bind address inside the container, and thus serves no purpose when running in a container

## Data Directory

```yaml
key: PISTON_DATA_DIRECTORY
default: /piston
```

Absolute path to piston related data, including packages and job contexts.

<!-- prettier-ignore -->
!!! warning
    Changing this value is not recommended.

    Some packages require absolute paths on disk at build time.
    Due to this, some packages may break when changing this parameter.

## Runner GID/UID range

```yaml
key:
    - PISTON_RUNNER_UID_MIN
    - PISTON_RUNNER_UID_MAX
    - PISTON_RUNNER_GID_MIN
    - PISTON_RUNNER_GID_MAX
default:
    - 1001
    - 1500
    - 1001
    - 1500
```

UID and GID ranges to use when executing jobs.

<!-- prettier-ignore -->
!!! warning
    Changing this value is not recommended.

    The piston container creates 500 users and groups by default, and reserves user/group 1000 for running the API.
    Any processes run by these users will be killed when cleaning up a job.

## Disable Networking

```yaml
key: PISTON_DISABLE_NETWORKING
default: true
```

Disallows access to `socket` syscalls, effectively disabling networking for jobs run by piston.

## Max Process Count

```yaml
key: PISTON_MAX_PROCESS_COUNT
default: 64
```

Maximum number of processes allowed to to have open for a job.

Resists against exhausting the process table, causing a full system lockup.

## Output Max Size

```yaml
key: PISTON_OUTPUT_MAX_SIZE
default: 1024
```

Maximum size of stdio buffers for each job.

Resist against run-away output which could lead to memory exhaustion.

## Max Open Files

```yaml
key: PISTON_MAX_OPEN_FILES
default: 64
```

Maximum number of open files at a given time by a job.

Resists against writing many smaller files to exhaust inodes.

## Max File Size

```yaml
key: PISTON_MAX_FILE_SIZE
default: 10000000 #10MB
```

Maximum size for a singular file written to disk.

Resists against large file writes to exhaust disk space.

## Compile/Run timeouts

```yaml
key:
  - PISTON_COMPILE_TIMEOUT
default: 10000

key:
  - PISTON_RUN_TIMEOUT
default: 3000
```

The maximum time that is allowed to be taken by a stage in milliseconds.
Use -1 for unlimited time.

## Compile/Run memory limits

```yaml
key:
    - PISTON_COMPILE_MEMORY_LIMIT
    - PISTON_RUN_MEMORY_LIMIT
default: -1
```

Maximum memory allowed by a stage in bytes.
Use -1 for unlimited memory usage.

Useful for running memory-limited contests.

## Repository URL

```yaml
key: PISTON_REPO_URL
default: https://github.com/luj8n/piston/releases/download/pkgs/index
```

URL for repository index, where packages will be downloaded from.

## Maximum Concurrent Jobs

```yaml
key: PISTON_MAX_CONCURRENT_JOBS
default: 64
```

Maximum number of jobs to run concurrently.

## Limit overrides

```yaml
key: PISTON_LIMIT_OVERRIDES
default: {}
```

Per-language overrides/exceptions for the each of `max_process_count`, `max_open_files`, `max_file_size`,
`compile_memory_limit`, `run_memory_limit`, `compile_timeout`, `run_timeout`, `output_max_size`. Defined as follows:

```
PISTON_LIMIT_OVERRIDES={"c++":{"max_process_count":128}}
```

This will give `c++` a max_process_count of 128 regardless of the configuration.


api/src/config.js
const fss = require('fs');
const Logger = require('logplease');
const logger = Logger.create('config');

function parse_overrides(overrides) {
    try {
        return JSON.parse(overrides);
    } catch (e) {
        return null;
    }
}

function validate_overrides(overrides, options) {
    for (const language in overrides) {
        for (const key in overrides[language]) {
            if (
                ![
                    'max_process_count',
                    'max_open_files',
                    'max_file_size',
                    'compile_memory_limit',
                    'run_memory_limit',
                    'compile_timeout',
                    'run_timeout',
                    'output_max_size',
                ].includes(key)
            ) {
                logger.error(`Invalid overridden option: ${key}`);
                return false;
            }
            const option = options.find(o => o.key === key);
            const parser = option.parser;
            const raw = overrides[language][key];
            const value = parser(raw);
            const validators = option.validators;
            for (const validator of validators) {
                const response = validator(value, raw);
                if (response !== true) {
                    logger.error(
                        `Failed to validate overridden option: ${key}`,
                        response
                    );
                    return false;
                }
            }
            overrides[language][key] = value;
        }
        // Modifies the reference
        options[
            options.index_of(options.find(o => o.key === 'limit_overrides'))
        ] = overrides;
    }
    return true;
}

const options = [
    {
        key: 'log_level',
        desc: 'Level of data to log',
        default: 'INFO',
        options: Object.values(Logger.LogLevels),
        validators: [
            x =>
                Object.values(Logger.LogLevels).includes(x) ||
                `Log level ${x} does not exist`,
        ],
    },
    {
        key: 'bind_address',
        desc: 'Address to bind REST API on',
        default: '0.0.0.0:2000',
        validators: [],
    },
    {
        key: 'data_directory',
        desc: 'Absolute path to store all piston related data at',
        default: '/piston',
        validators: [
            x => fss.exists_sync(x) || `Directory ${x} does not exist`,
        ],
    },
    {
        key: 'runner_uid_min',
        desc: 'Minimum uid to use for runner',
        default: 1001,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'runner_uid_max',
        desc: 'Maximum uid to use for runner',
        default: 1500,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'runner_gid_min',
        desc: 'Minimum gid to use for runner',
        default: 1001,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'runner_gid_max',
        desc: 'Maximum gid to use for runner',
        default: 1500,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'disable_networking',
        desc: 'Set to true to disable networking',
        default: true,
        parser: x => x === 'true',
        validators: [x => typeof x === 'boolean' || `${x} is not a boolean`],
    },
    {
        key: 'output_max_size',
        desc: 'Max size of each stdio buffer',
        default: 1024,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'max_process_count',
        desc: 'Max number of processes per job',
        default: 64,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'max_open_files',
        desc: 'Max number of open files per job',
        default: 2048,
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'max_file_size',
        desc: 'Max file size in bytes for a file',
        default: 10000000, //10MB
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'compile_timeout',
        desc: 'Max time allowed for compile stage in milliseconds',
        default: 10000, // 10 seconds
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'run_timeout',
        desc: 'Max time allowed for run stage in milliseconds',
        default: 3000, // 3 seconds
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'compile_memory_limit',
        desc: 'Max memory usage for compile stage in bytes (set to -1 for no limit)',
        default: -1, // no limit
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'run_memory_limit',
        desc: 'Max memory usage for run stage in bytes (set to -1 for no limit)',
        default: -1, // no limit
        parser: parse_int,
        validators: [(x, raw) => !is_nan(x) || `${raw} is not a number`],
    },
    {
        key: 'repo_url',
        desc: 'URL of repo index',
        default: 'https://github.com/luj8n/piston/releases/download/pkgs/index',
        validators: [],
    },
    {
        key: 'max_concurrent_jobs',
        desc: 'Maximum number of concurrent jobs to run at one time',
        default: 64,
        parser: parse_int,
        validators: [x => x > 0 || `${x} cannot be negative`],
    },
    {
        key: 'limit_overrides',
        desc: 'Per-language exceptions in JSON format for each of:\
        max_process_count, max_open_files, max_file_size, compile_memory_limit,\
        run_memory_limit, compile_timeout, run_timeout, output_max_size',
        default: {},
        parser: parse_overrides,
        validators: [
            x => !!x || `Invalid JSON format for the overrides\n${x}`,
            (overrides, _, options) =>
                validate_overrides(overrides, options) ||
                `Failed to validate the overrides`,
        ],
    },
];

logger.info(`Loading Configuration from environment`);

let errored = false;

let config = {};

options.forEach(option => {
    const env_key = 'PISTON_' + option.key.to_upper_case();

    const parser = option.parser || (x => x);

    const env_val = process.env[env_key];

    const parsed_val = parser(env_val);

    const value = env_val === undefined ? option.default : parsed_val;

    option.validators.for_each(validator => {
        let response = null;
        if (env_val) response = validator(parsed_val, env_val, options);
        else response = validator(value, value, options);

        if (response !== true) {
            errored = true;
            logger.error(
                `Config option ${option.key} failed validation:`,
                response
            );
            return;
        }
    });

    config[option.key] = value;
});

if (errored) {
    process.exit(1);
}

logger.info('Configuration successfully loaded');

module.exports = config;
